src/ts-ast.ts

```diff
@@ -325,6 +400,47 @@ export class UnionType {
     this.members = members;
   }
 
+  walk(cb: (node: Node) => void) {
+    this.members.map((m) => m.walk(cb));
+    cb(this);
+  }
+
+  /**
+   * Simplify this union type:
+   *
+   * 1) Flatten nested unions (`foo|(bar|baz)` -> `foo|bar|baz`).
+   * 2) De-duplicate identical members (`foo|bar|foo` -> `foo|bar`).
+   */
+  simplify() {
+    const flattened = [];
+    for (const m of this.members) {
+      if (m.kind === 'union') {
+        // Note we are not recursing here, because we assume we're being called
+        // via a depth-first walk, so any union members have already been
+        // simplified.
+        flattened.push(...m.members);
+      } else {
+        flattened.push(m);
+      }
+    }
+
+    // TODO This only de-dupes Name types. We should de-dupe Arrays and
+    // Functions too.
+    const deduped = [];
+    const names = new Set();
+    for (const m of flattened) {
+      if (m.kind === 'name') {
+        if (names.has(m.name)) {
+          continue;
+        }
+        names.add(m.name);
+      }
+      deduped.push(m);
+    }
```

Comment:
Feature request: move `null` and `undefined` to the end of the array. `DomEvent|null|DomEvent[]` is harder to read

Reply:
Good idea. Added TODO for now.